# Importing necessary libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# Importing necessary csv files

In [2]:
df_ratings = pd.read_csv("D:/Spyder/Data-Envelopment-Analysis-Excel/Data/MOVIE PROJECT/ml-25m/ml-25m/ratings.csv")
df_movies = pd.read_csv("D:/Spyder/Data-Envelopment-Analysis-Excel/Data/MOVIE PROJECT/ml-25m/ml-25m/movies.csv", encoding  = 'unicode_escape') #contains the title
df_tags= pd.read_csv("D:/Spyder/Data-Envelopment-Analysis-Excel/Data/MOVIE PROJECT/ml-25m/ml-25m/tags.csv") 
df_genome_scores = pd.read_csv("D:/Spyder/Data-Envelopment-Analysis-Excel/Data/MOVIE PROJECT/ml-25m/ml-25m/genome-scores.csv") 
df_links = pd.read_csv("D:/Spyder/Data-Envelopment-Analysis-Excel/Data/MOVIE PROJECT/ml-25m/ml-25m/links.csv")
df_genome_tags = pd.read_csv("D:/Spyder/Data-Envelopment-Analysis-Excel/Data/MOVIE PROJECT/ml-25m/ml-25m/genome-tags.csv")

# Checking each of them

### df_movies

In [3]:
df_movies.head()

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

### df_links

In [4]:
df_links.head()

movieId  imdbId   tmdbId
0        1  114709    862.0
1        2  113497   8844.0
2        3  113228  15602.0
3        4  114885  31357.0
4        5  113041  11862.0

### df_ratings

In [5]:
df_ratings.head()

userId  movieId  rating   timestamp
0       1      296     5.0  1147880044
1       1      306     3.5  1147868817
2       1      307     5.0  1147868828
3       1      665     5.0  1147878820
4       1      899     3.5  1147868510

In [6]:
df_ratings.shape

(25000095, 4)

In [7]:
df_ratings['userId'].value_counts().shape

(162541,)

#### out of 25000095 ratings done, 162541 are the unique raters. this is till a big number, so now we will narrow our search by removing all those users, who hadn't rated at least 500 movies

In [8]:
x = df_ratings['userId'].value_counts() > 500

In [9]:
y = x[x].index

In [10]:
y.shape

(9663,)

In [11]:
rating = df_ratings[df_ratings['userId'].isin(y)]

In [12]:
rating.shape

(8935061, 4)

In [13]:
rating

userId  movieId  rating   timestamp
254            3        1     4.0  1439472215
255            3       29     4.5  1484754967
256            3       32     4.5  1439474635
257            3       50     5.0  1439474391
258            3      111     4.0  1484753849
...          ...      ...     ...         ...
24996936  162519     4748     2.0  1000947348
24996937  162519     4789     1.0  1000946127
24996938  162519     4855     5.0  1001884147
24996939  162519     4857     5.0  1001886067
24996940  162519     5060     5.0  1000948162

[8935061 rows x 4 columns]

### Now merging df_movies and rating

In [14]:
movie_details = df_movies.merge(rating, on = 'movieId')

In [15]:
movie_details.head()

movieId             title                                       genres  \
0        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
1        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
2        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
3        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
4        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   

   userId  rating   timestamp  
0       3     4.0  1439472215  
1      12     4.0  1167582601  
2     120     5.0   956264593  
3     166     4.0   942014736  
4     171     4.5  1074594380

In [16]:
movie_details.shape

(8935061, 6)

### Lets remove timestamp column as we don't need it

In [17]:
movie_details.drop(columns=['timestamp'], inplace = True)

In [18]:
movie_details.shape

(8935061, 5)

In [19]:
movie_details.head()

movieId             title                                       genres  \
0        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
1        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
2        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
3        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
4        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   

   userId  rating  
0       3     4.0  
1      12     4.0  
2     120     5.0  
3     166     4.0  
4     171     4.5

#### getting the info that which movie is rated how many number of times

In [20]:
number_rating = movie_details.groupby('title')['rating'].count().reset_index()

In [21]:
number_rating.rename(columns = {'rating' : 'number of rating'}, inplace = True)

In [22]:
number_rating.head()

title  number of rating
0  "BLOW THE NIGHT!" Let's Spend the Night Togeth...                 1
1                   "Great Performances" Cats (1998)                42
2                         #1 Cheerleader Camp (2010)                 7
3                                   #Captured (2017)                 2
4                            #Female Pleasure (2018)                 2

### Now merging movie_details and number_rating

In [23]:
df_tillnow = movie_details.merge(number_rating, on = 'title')

In [24]:
df_tillnow.shape

(8935061, 6)

In [25]:
df_tillnow.head()

movieId             title                                       genres  \
0        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
1        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
2        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
3        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
4        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   

   userId  rating  number of rating  
0       3     4.0              8078  
1      12     4.0              8078  
2     120     5.0              8078  
3     166     4.0              8078  
4     171     4.5              8078

#### by setting the threshold of 50 in number_of_rating so as to more narrow down our search space

In [26]:
df_tillnow1 = df_tillnow[df_tillnow['number of rating']>=50] #taking only the valuable one

In [27]:
df_tillnow1.shape

(8596657, 6)

In [28]:
df_tillnow1.head()

movieId             title                                       genres  \
0        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
1        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
2        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
3        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
4        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   

   userId  rating  number of rating  
0       3     4.0              8078  
1      12     4.0              8078  
2     120     5.0              8078  
3     166     4.0              8078  
4     171     4.5              8078

#### Now we don't need number_of_rating column

In [29]:
df_tillnow1.drop(columns=['number of rating'], inplace = True)

C:\Users\SHIVLEON\Anaconda3\lib\site-packages\pandas\core\frame.py:4315: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [30]:
df_tillnow1.head()

movieId             title                                       genres  \
0        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
1        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
2        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
3        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
4        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   

   userId  rating  
0       3     4.0  
1      12     4.0  
2     120     5.0  
3     166     4.0  
4     171     4.5

In [31]:
df_tillnow1['rating'] = df_tillnow1['rating'].astype(float)

C:\Users\SHIVLEON\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [32]:
df_tillnow1.head()

movieId             title                                       genres  \
0        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
1        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
2        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
3        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   
4        1  Toy Story (1995)  Adventure|Animation|Children|Comedy|Fantasy   

   userId  rating  
0       3     4.0  
1      12     4.0  
2     120     5.0  
3     166     4.0  
4     171     4.5

### Lets try to see each movie average rating

In [33]:
average_rating_df = df_tillnow1.groupby('title')['rating'].mean().reset_index()

In [34]:
average_rating_df

title    rating
0                                       '71 (2014)  3.568889
1                           'Round Midnight (1986)  3.701031
2                              'Salem's Lot (2004)  3.123188
3                        'Til There Was You (1997)  2.491124
4                               'burbs, The (1989)  3.050789
...                                            ...       ...
11711                                   xXx (2002)  2.662130
11712            xXx: Return of Xander Cage (2017)  2.638298
11713               xXx: State of the Union (2005)  2.224012
11714                       ¡Three Amigos! (1986)  3.050126
11715  À nous la liberté (Freedom for Us) (1931)  3.528226

[11716 rows x 2 columns]

### Now merging df_movies and average_rating_df

In [35]:
average_rating_movies = df_movies.merge(average_rating_df, on = 'title') 

In [36]:
average_rating_movies

movieId                                   title  \
0            1                        Toy Story (1995)   
1            2                          Jumanji (1995)   
2            3                 Grumpier Old Men (1995)   
3            4                Waiting to Exhale (1995)   
4            5      Father of the Bride Part II (1995)   
...        ...                                     ...   
11744   204352                         Ad Astra (2019)   
11745   204542                  It: Chapter Two (2019)   
11746   204698                            Joker (2019)   
11747   205383  El Camino: A Breaking Bad Movie (2019)   
11748   206499     Between Two Ferns: The Movie (2019)   

                                            genres    rating  
0      Adventure|Animation|Children|Comedy|Fantasy  3.828113  
1                       Adventure|Children|Fantasy  3.059255  
2                                   Comedy|Romance  2.848990  
3                             Comedy|Drama|Romance  2.547771  
4                                           Comedy  2.708668  
...                                            ...       ...  
11744      Adventure|Drama|Mystery|Sci-Fi|Thriller  3.364035  
11745                                       Horror  3.108108  
11746                         Crime|Drama|Thriller  3.968843  
11747                         Crime|Drama|Thriller  3.520000  
11748                                       Comedy  3.000000  

[11749 rows x 4 columns]

In [37]:
import re
def changes_in_movie_name(data):
    data = data.lower()
    data = re.sub(r"\([^()]*\)", "", data)
    data = data.strip()
    return data

In [38]:
average_rating_movies['title'] = average_rating_movies['title'].apply(lambda x : changes_in_movie_name(x))

In [39]:
average_rating_movies

movieId                            title  \
0            1                        toy story   
1            2                          jumanji   
2            3                 grumpier old men   
3            4                waiting to exhale   
4            5      father of the bride part ii   
...        ...                              ...   
11744   204352                         ad astra   
11745   204542                  it: chapter two   
11746   204698                            joker   
11747   205383  el camino: a breaking bad movie   
11748   206499     between two ferns: the movie   

                                            genres    rating  
0      Adventure|Animation|Children|Comedy|Fantasy  3.828113  
1                       Adventure|Children|Fantasy  3.059255  
2                                   Comedy|Romance  2.848990  
3                             Comedy|Drama|Romance  2.547771  
4                                           Comedy  2.708668  
...                                            ...       ...  
11744      Adventure|Drama|Mystery|Sci-Fi|Thriller  3.364035  
11745                                       Horror  3.108108  
11746                         Crime|Drama|Thriller  3.968843  
11747                         Crime|Drama|Thriller  3.520000  
11748                                       Comedy  3.000000  

[11749 rows x 4 columns]

#### Now we are gonna convert genres into each column using dummy

In [40]:
dummy_genre = average_rating_movies['genres'].str.get_dummies(sep = "|")
content_based_df = pd.concat([average_rating_movies, dummy_genre], axis = 1)

In [41]:
content_based_df

movieId                            title  \
0            1                        toy story   
1            2                          jumanji   
2            3                 grumpier old men   
3            4                waiting to exhale   
4            5      father of the bride part ii   
...        ...                              ...   
11744   204352                         ad astra   
11745   204542                  it: chapter two   
11746   204698                            joker   
11747   205383  el camino: a breaking bad movie   
11748   206499     between two ferns: the movie   

                                            genres    rating  \
0      Adventure|Animation|Children|Comedy|Fantasy  3.828113   
1                       Adventure|Children|Fantasy  3.059255   
2                                   Comedy|Romance  2.848990   
3                             Comedy|Drama|Romance  2.547771   
4                                           Comedy  2.708668   
...                                            ...       ...   
11744      Adventure|Drama|Mystery|Sci-Fi|Thriller  3.364035   
11745                                       Horror  3.108108   
11746                         Crime|Drama|Thriller  3.968843   
11747                         Crime|Drama|Thriller  3.520000   
11748                                       Comedy  3.000000   

       (no genres listed)  Action  Adventure  Animation  Children  Comedy  \
0                       0       0          1          1         1       1   
1                       0       0          1          0         1       0   
2                       0       0          0          0         0       1   
3                       0       0          0          0         0       1   
4                       0       0          0          0         0       1   
...                   ...     ...        ...        ...       ...     ...   
11744                   0       0          1          0         0       0   
11745                   0       0          0          0         0       0   
11746                   0       0          0          0         0       0   
11747                   0       0          0          0         0       0   
11748                   0       0          0          0         0       1   

       ...  Film-Noir  Horror  IMAX  Musical  Mystery  Romance  Sci-Fi  \
0      ...          0       0     0        0        0        0       0   
1      ...          0       0     0        0        0        0       0   
2      ...          0       0     0        0        0        1       0   
3      ...          0       0     0        0        0        1       0   
4      ...          0       0     0        0        0        0       0   
...    ...        ...     ...   ...      ...      ...      ...     ...   
11744  ...          0       0     0        0        1        0       1   
11745  ...          0       1     0        0        0        0       0   
11746  ...          0       0     0        0        0        0       0   
11747  ...          0       0     0        0        0        0       0   
11748  ...          0       0     0        0        0        0       0   

       Thriller  War  Western  
0             0    0        0  
1             0    0        0  
2             0    0        0  
3             0    0        0  
4             0    0        0  
...         ...  ...      ...  
11744         1    0        0  
11745         0    0        0  
11746         1    0        0  
11747         1    0        0  
11748         0    0        0  

[11749 rows x 24 columns]

#### Now droping the main genre column, as we have the dummy one

In [42]:
content_based_df_main = content_based_df.drop(["genres"], axis = 1)

In [43]:
content_based_df_main

movieId                            title    rating  (no genres listed)  \
0            1                        toy story  3.828113                   0   
1            2                          jumanji  3.059255                   0   
2            3                 grumpier old men  2.848990                   0   
3            4                waiting to exhale  2.547771                   0   
4            5      father of the bride part ii  2.708668                   0   
...        ...                              ...       ...                 ...   
11744   204352                         ad astra  3.364035                   0   
11745   204542                  it: chapter two  3.108108                   0   
11746   204698                            joker  3.968843                   0   
11747   205383  el camino: a breaking bad movie  3.520000                   0   
11748   206499     between two ferns: the movie  3.000000                   0   

       Action  Adventure  Animation  Children  Comedy  Crime  ...  Film-Noir  \
0           0          1          1         1       1      0  ...          0   
1           0          1          0         1       0      0  ...          0   
2           0          0          0         0       1      0  ...          0   
3           0          0          0         0       1      0  ...          0   
4           0          0          0         0       1      0  ...          0   
...       ...        ...        ...       ...     ...    ...  ...        ...   
11744       0          1          0         0       0      0  ...          0   
11745       0          0          0         0       0      0  ...          0   
11746       0          0          0         0       0      1  ...          0   
11747       0          0          0         0       0      1  ...          0   
11748       0          0          0         0       1      0  ...          0   

       Horror  IMAX  Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0           0     0        0        0        0       0         0    0        0  
1           0     0        0        0        0       0         0    0        0  
2           0     0        0        0        1       0         0    0        0  
3           0     0        0        0        1       0         0    0        0  
4           0     0        0        0        0       0         0    0        0  
...       ...   ...      ...      ...      ...     ...       ...  ...      ...  
11744       0     0        0        1        0       1         1    0        0  
11745       1     0        0        0        0       0         0    0        0  
11746       0     0        0        0        0       0         1    0        0  
11747       0     0        0        0        0       0         1    0        0  
11748       0     0        0        0        0       0         0    0        0  

[11749 rows x 23 columns]

In [44]:
content_based_df_main.shape

(11749, 23)

## Our main dataframes to work with:::
### 1 - df_tillnow1
### 2 - content_based_df_main

#### Changing the names of movie in df_tillnow1 as well

In [45]:
df_tillnow1['title'] = df_tillnow1['title'].apply(lambda x : changes_in_movie_name(x))

C:\Users\SHIVLEON\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [46]:
df_tillnow1

movieId                         title  \
0              1                     toy story   
1              1                     toy story   
2              1                     toy story   
3              1                     toy story   
4              1                     toy story   
...          ...                           ...   
8933940   206499  between two ferns: the movie   
8933941   206499  between two ferns: the movie   
8933942   206499  between two ferns: the movie   
8933943   206499  between two ferns: the movie   
8933944   206499  between two ferns: the movie   

                                              genres  userId  rating  
0        Adventure|Animation|Children|Comedy|Fantasy       3     4.0  
1        Adventure|Animation|Children|Comedy|Fantasy      12     4.0  
2        Adventure|Animation|Children|Comedy|Fantasy     120     5.0  
3        Adventure|Animation|Children|Comedy|Fantasy     166     4.0  
4        Adventure|Animation|Children|Comedy|Fantasy     171     4.5  
...                                              ...     ...     ...  
8933940                                       Comedy  145866     3.5  
8933941                                       Comedy  151072     3.0  
8933942                                       Comedy  153802     3.0  
8933943                                       Comedy  154858     4.0  
8933944                                       Comedy  159927     3.5  

[8596657 rows x 5 columns]

#### Now making similarity matrix, i.e., User-Item Similarity Matrix

In [47]:
movie_similarity_matrix = df_tillnow1.pivot_table(columns = "title", index = "userId", values = "rating")

In [48]:
movie_similarity_matrix

title   '71  'burbs, the  'night mother  'round midnight  'salem's lot  \
userId                                                                   
3       NaN          NaN            NaN              NaN           NaN   
12      NaN          NaN            NaN              NaN           NaN   
72      NaN          3.0            NaN              NaN           NaN   
80      NaN          NaN            NaN              NaN           NaN   
120     NaN          NaN            NaN              NaN           NaN   
...     ...          ...            ...              ...           ...   
162484  NaN          NaN            NaN              NaN           NaN   
162495  NaN          3.5            NaN              NaN           NaN   
162508  NaN          NaN            NaN              NaN           NaN   
162516  NaN          NaN            NaN              2.5           NaN   
162519  NaN          NaN            NaN              NaN           NaN   

title   'til there was you  *batteries not included  ...and justice for all  \
userId                                                                        
3                      NaN                      NaN                     NaN   
12                     NaN                      NaN                     NaN   
72                     NaN                      NaN                     NaN   
80                     NaN                      NaN                     NaN   
120                    NaN                      NaN                     NaN   
...                    ...                      ...                     ...   
162484                 NaN                      NaN                     NaN   
162495                 NaN                      3.5                     NaN   
162508                 NaN                      NaN                     NaN   
162516                 NaN                      3.0                     3.0   
162519                 NaN                      NaN                     NaN   

title    10  10 cloverfield lane  ...  zookeeper  zoolander  zoolander 2  \
userId                            ...                                      
3       NaN                  NaN  ...        NaN        3.5          NaN   
12      NaN                  NaN  ...        NaN        NaN          NaN   
72      NaN                  NaN  ...        NaN        NaN          NaN   
80      NaN                  NaN  ...        NaN        NaN          NaN   
120     NaN                  NaN  ...        NaN        NaN          NaN   
...     ...                  ...  ...        ...        ...          ...   
162484  NaN                  NaN  ...        NaN        3.5          NaN   
162495  NaN                  NaN  ...        NaN        3.5          NaN   
162508  NaN                  NaN  ...        NaN        NaN          NaN   
162516  3.0                  3.5  ...        NaN        3.0          NaN   
162519  NaN                  NaN  ...        NaN        NaN          NaN   

title   zoom  zootopia  zorba the greek  zorro, the gay blade  zulu  \
userId                                                                
3        NaN       NaN              NaN                   NaN   NaN   
12       NaN       NaN              NaN                   NaN   NaN   
72       NaN       NaN              NaN                   NaN   NaN   
80       NaN       NaN              NaN                   NaN   NaN   
120      NaN       NaN              NaN                   NaN   NaN   
...      ...       ...              ...                   ...   ...   
162484   NaN       NaN              NaN                   NaN   NaN   
162495   NaN       NaN              NaN                   NaN   NaN   
162508   NaN       NaN              NaN                   NaN   NaN   
162516   NaN       4.0              NaN                   1.5   3.0   
162519   NaN       NaN              NaN                   NaN   NaN   

title   â¡three amigos!  〠nous la libertã©  
userId                                        
3 

In [49]:
movie_similarity_matrix.shape

(9663, 11274)

In [50]:
movie_similarity_matrix.fillna(0, inplace = True)

In [51]:
type(movie_similarity_matrix)

pandas.core.frame.DataFrame

In [52]:
movie_similarity_matrix_copy = movie_similarity_matrix.copy(deep = True)

In [53]:
movie_similarity_matrix_copy

title   '71  'burbs, the  'night mother  'round midnight  'salem's lot  \
userId                                                                   
3       0.0          0.0            0.0              0.0           0.0   
12      0.0          0.0            0.0              0.0           0.0   
72      0.0          3.0            0.0              0.0           0.0   
80      0.0          0.0            0.0              0.0           0.0   
120     0.0          0.0            0.0              0.0           0.0   
...     ...          ...            ...              ...           ...   
162484  0.0          0.0            0.0              0.0           0.0   
162495  0.0          3.5            0.0              0.0           0.0   
162508  0.0          0.0            0.0              0.0           0.0   
162516  0.0          0.0            0.0              2.5           0.0   
162519  0.0          0.0            0.0              0.0           0.0   

title   'til there was you  *batteries not included  ...and justice for all  \
userId                                                                        
3                      0.0                      0.0                     0.0   
12                     0.0                      0.0                     0.0   
72                     0.0                      0.0                     0.0   
80                     0.0                      0.0                     0.0   
120                    0.0                      0.0                     0.0   
...                    ...                      ...                     ...   
162484                 0.0                      0.0                     0.0   
162495                 0.0                      3.5                     0.0   
162508                 0.0                      0.0                     0.0   
162516                 0.0                      3.0                     3.0   
162519                 0.0                      0.0                     0.0   

title    10  10 cloverfield lane  ...  zookeeper  zoolander  zoolander 2  \
userId                            ...                                      
3       0.0                  0.0  ...        0.0        3.5          0.0   
12      0.0                  0.0  ...        0.0        0.0          0.0   
72      0.0                  0.0  ...        0.0        0.0          0.0   
80      0.0                  0.0  ...        0.0        0.0          0.0   
120     0.0                  0.0  ...        0.0        0.0          0.0   
...     ...                  ...  ...        ...        ...          ...   
162484  0.0                  0.0  ...        0.0        3.5          0.0   
162495  0.0                  0.0  ...        0.0        3.5          0.0   
162508  0.0                  0.0  ...        0.0        0.0          0.0   
162516  3.0                  3.5  ...        0.0        3.0          0.0   
162519  0.0                  0.0  ...        0.0        0.0          0.0   

title   zoom  zootopia  zorba the greek  zorro, the gay blade  zulu  \
userId                                                                
3        0.0       0.0              0.0                   0.0   0.0   
12       0.0       0.0              0.0                   0.0   0.0   
72       0.0       0.0              0.0                   0.0   0.0   
80       0.0       0.0              0.0                   0.0   0.0   
120      0.0       0.0              0.0                   0.0   0.0   
...      ...       ...              ...                   ...   ...   
162484   0.0       0.0              0.0                   0.0   0.0   
162495   0.0       0.0              0.0                   0.0   0.0   
162508   0.0       0.0              0.0                   0.0   0.0   
162516   0.0       4.0              0.0                   1.5   3.0   
162519   0.0       0.0              0.0                   0.0   0.0   

title   â¡three amigos!  〠nous la libertã©  
userId                                        
3 

In [54]:
movie_list = ["toy story", "heat", "goldeneye", "nixon"]

In [55]:
df = pd.DataFrame({'Region':['West','North','South'],
                   'Company':['Costco','Walmart','Home Depot'],
                   'Product':['Dinner Set','Grocery','Gardening tools'],
                   'Month':['September','July','February'],
                   'Sales':[2500,3096,8795]})

In [56]:
data = [{'Region' : 'East', 'Product' : 'Condom'}]

In [57]:
df.append(data)

Region     Company          Product      Month   Sales
0   West      Costco       Dinner Set  September  2500.0
1  North     Walmart          Grocery       July  3096.0
2  South  Home Depot  Gardening tools   February  8795.0
0   East         NaN           Condom        NaN     NaN

In [58]:
data = [{'Region' : 'East', 'Product' : 'Condoms'}]

In [59]:
df.append(data)

Region     Company          Product      Month   Sales
0   West      Costco       Dinner Set  September  2500.0
1  North     Walmart          Grocery       July  3096.0
2  South  Home Depot  Gardening tools   February  8795.0
0   East         NaN          Condoms        NaN     NaN

In [60]:
data = [{'Region' : 'South-East', 'Product' : 'Lube'}]

In [61]:
df.append(data)

Region     Company          Product      Month   Sales
0        West      Costco       Dinner Set  September  2500.0
1       North     Walmart          Grocery       July  3096.0
2       South  Home Depot  Gardening tools   February  8795.0
0  South-East         NaN             Lube        NaN     NaN

In [62]:
movie_list += ['zootopia']

In [63]:
movie_list

['toy story', 'heat', 'goldeneye', 'nixon', 'zootopia']

In [64]:
movie_rating = [4.5, 1.2, 2.6, 4.3, 4.9]

In [65]:
dic_watched = {}
for key in movie_list:
    for value in movie_rating:
        dic_watched[key] = value
        movie_rating.remove(value)
        break
    

In [66]:
dic_watched

{'toy story': 4.5,
 'heat': 1.2,
 'goldeneye': 2.6,
 'nixon': 4.3,
 'zootopia': 4.9}

In [67]:
movie_similarity_matrix_copy = movie_similarity_matrix_copy.append(dic_watched, ignore_index=True)

In [68]:
dic_watched['zoom'] = 4.5

In [69]:
movie_similarity_matrix_copy = movie_similarity_matrix_copy.append(pd.Series(dic_watched), ignore_index=True)

In [70]:
movie_similarity_matrix_copy.fillna(0, inplace = True)

In [71]:
movie_similarity_matrix_copy

'71  'burbs, the  'night mother  'round midnight  'salem's lot  \
0     0.0          0.0            0.0              0.0           0.0   
1     0.0          0.0            0.0              0.0           0.0   
2     0.0          3.0            0.0              0.0           0.0   
3     0.0          0.0            0.0              0.0           0.0   
4     0.0          0.0            0.0              0.0           0.0   
...   ...          ...            ...              ...           ...   
9660  0.0          0.0            0.0              0.0           0.0   
9661  0.0          0.0            0.0              2.5           0.0   
9662  0.0          0.0            0.0              0.0           0.0   
9663  0.0          0.0            0.0              0.0           0.0   
9664  0.0          0.0            0.0              0.0           0.0   

      'til there was you  *batteries not included  ...and justice for all  \
0                    0.0                      0.0                     0.0   
1                    0.0                      0.0                     0.0   
2                    0.0                      0.0                     0.0   
3                    0.0                      0.0                     0.0   
4                    0.0                      0.0                     0.0   
...                  ...                      ...                     ...   
9660                 0.0                      0.0                     0.0   
9661                 0.0                      3.0                     3.0   
9662                 0.0                      0.0                     0.0   
9663                 0.0                      0.0                     0.0   
9664                 0.0                      0.0                     0.0   

       10  10 cloverfield lane  ...  zookeeper  zoolander  zoolander 2  zoom  \
0     0.0                  0.0  ...        0.0        3.5          0.0   0.0   
1     0.0                  0.0  ...        0.0        0.0          0.0   0.0   
2     0.0                  0.0  ...        0.0        0.0          0.0   0.0   
3     0.0                  0.0  ...        0.0        0.0          0.0   0.0   
4     0.0                  0.0  ...        0.0        0.0          0.0   0.0   
...   ...                  ...  ...        ...        ...          ...   ...   
9660  0.0                  0.0  ...        0.0        0.0          0.0   0.0   
9661  3.0                  3.5  ...        0.0        3.0          0.0   0.0   
9662  0.0                  0.0  ...        0.0        0.0          0.0   0.0   
9663  0.0                  0.0  ...        0.0        0.0          0.0   0.0   
9664  0.0                  0.0  ...        0.0        0.0          0.0   4.5   

      zootopia  zorba the greek  zorro, the gay blade  zulu  â¡three amigos!  \
0          0.0              0.0                   0.0   0.0              0.0   
1          0.0              0.0                   0.0   0.0              3.5   
2          0.0              0.0                   0.0   0.0              2.0   
3          0.0              0.0                   0.0   0.0              0.0   
4          0.0              0.0                   0.0   0.0              4.0   
...        ...              ...                   ...   ...              ...   
9660       0.0              0.0                   0.0   0.0              0.0   
9661       4.0              0.0                   1.5   3.0              2.0   
9662       0.0              0.0                   0.0   0.0              0.0   
9663       4.9              0.0                   0.0   0.0              0.0   
9664       4.9              0.0                   0.0   0.0              0.0   

      〠nous la libert㩠 
0                     0.0  
1                     0.0  
2                     0.0  
3                     0.0  
4                     0.0  
...                   ...  
9660                  0.0  
9661                  0.0  
9662                  0.0  
9663          

In [72]:
from scipy.sparse import csr_matrix

In [73]:
movie_sparse = csr_matrix(movie_similarity_matrix_copy)

In [74]:
movie_sparse

<9665x11274 sparse matrix of type '<class 'numpy.float64'>'
	with 8527804 stored elements in Compressed Sparse Row format>

In [75]:
from scipy.sparse import csr_matrix
from sklearn.neighbors import NearestNeighbors
def recommendation_part1_CF(movie_list, movie_ratings):
    #making dictionary from movie list and rating
    dic_watched1 = {}
    for key in movie_list:
        for value in movie_rating:
            dic_watched1[key] = value
            movie_rating.remove(value)
            break
    
    #Making a copy of original movie similarity matrix, so that the original stays intact
    #Keeping deep = True, so that a separate copy is made
    movie_similarity_matrix_cf = movie_similarity_matrix.copy(deep = True)
    #print(movie_similarity_matrix_cf)
    
    #Now adding our dictionary as an row
    movie_similarity_matrix_cf = movie_similarity_matrix_cf.append(pd.DataFrame(dic_watched1), ignore_index=True)
    #print("\n\nBefore Nan\n\n\n",movie_similarity_matrix_cf)
    #print("\n\n", movie_similarity_matrix_cf.isnull().sum())
    
    #making nan to 0.0
    movie_similarity_matrix_cf.fillna(0, inplace = True)
    #print("\n\n\nAfter Nan\n\n\n")
    #print(movie_similarity_matrix_cf)
    
    #Creating sparse matrix on which our NearestNeighbors model will get fit
    movie_sparse_cf = csr_matrix(movie_similarity_matrix_cf)
    
    #Creating our NearestNeighbors model
    modelr = NearestNeighbors(n_neighbors = 11, algorithm = 'brute', metric = 'euclidean')
    #Fitting our model
    modelr.fit(movie_sparse_cf)
    
    #lets get the similar users
    distance, similar_users = modelr.kneighbors(movie_similarity_matrix_cf.iloc[-1].values.reshape(1, -1))
    
    #print(distance, "\n", similar_users)
    
    returning_similar_users = list()
    for i in range(1,len(similar_users[0])):
        returning_similar_users.append(similar_users[0][i])
    #print(returning_similar_users)
    
    similar_users_df = movie_similarity_matrix_cf.iloc[similar_users[0][1]].to_frame()
    similar_users_df.columns = ['ratings']
    for i in range(2, len(similar_users)):
        subdf = movie_similarity_matrix_cf.iloc[similar_users[0][i]].to_frame()
        subdf.columns = ['ratings']
        similar_users_df = similar_users_df.add(subdf, fill_value = 0)
        
    similar_users_df = similar_users_df[similar_users_df['ratings'] > 1]
    #print("\n\n\n", similar_users_df.head(15))
    
    
    #getting movies watched by similar users (contain names of movies watched by our user) 
    list_of_movies_by_similar_users = similar_users_df.index.tolist()
    #print(list_of_movies_by_similar_users, "\n\n\n")
    #list_of_movies_ratings_ = similar_users_df['ratings'].tolist()
    
    #remvoing already watched movies
    list_of_movies_by_similar_users = list(set(list_of_movies_by_similar_users) - set(movie_list))
    list_of_movies_ratings_ = similar_users_df.loc[list_of_movies_by_similar_users, 'ratings'].tolist();
    #print(len(list_of_movies_by_similar_users), "\nw", len(list_of_movies_ratings_))
    
    recommendation_part2_CB_and_Recommendation(movie_list, movie_ratings, list_of_movies_by_similar_users, list_of_movies_ratings_)

In [76]:
def recommendation_part2_CB_and_Recommendation(movie_list, movie_ratings, list_of_movies_by_similar_users, list_of_movies_ratings_):
    datauser = content_based_df_main.loc[content_based_df_main['title'].isin(movie_list)] #already watched movie
    datarecommendation = content_based_df_main.loc[content_based_df_main['title'].isin(set(list_of_movies_by_similar_users))] #from where the recommendation will come
    #print("\n\n\nDataUser\n\n")
    #print(datauser.shape)
    #print("\n\n\nDataRecommendation\n\n")
    #print(datarecommendation.shape)
    #print("\n\n", datauser.isnull().sum())  #no null
    #print("\n\n", datarecommendation.isnull().sum()) #no null
    
    #Replacing the rating in datauser (as its an average rating prior) with the movie_ratings we took as an arguement
    #print(len(list_of_movies_by_similar_users), " " , len(datarecommendation['title'].tolist()))
    datauser.loc[:,'rating'] = movie_ratings
    #datarecommendation.loc[:, 'rating'] = list_of_movies_ratings_
    #print("\n\n\nDataUser After Rating Updation\n\n")
    #print(datauser)
    
    #print(list_of_movies_by_similar_users, "\n\n")
    #print(datarecommendation['title'].tolist())
    #print(len(list_of_movies_by_similar_users), " " , len(datarecommendation['title'].tolist()))
    
    #Now taking out rating specifically for multiplication purpose
    list_user_rating = np.array(datauser['rating'])
    #print("\n\n", np.isnan(np.sum(list_user_rating))) #no null
    
    #making weighted matrix for datauser which will be multiplied by list_user_rating
    datauser_genre_matrix = datauser.drop(['movieId', 'title', 'rating'], axis=1)
    #print(datauser_genre_matrix)
    #print("\n\n", datauser_genre_matrix.isnull().sum()) #no nan
    
    list_user_rating = list_user_rating.reshape(len(list_user_rating), 1) #reshaping the matrix so that multiplication can happen
    #print(list_user_rating)
    
    #changing our datauser_genre_matrix into numpy array so that multiplication is possible
    data_user_genre_matrix_numpy = np.array(datauser_genre_matrix)
    #print(data_user_genre_matrix_numpy)
    
    #now multiplying our rating with our numpy matrix (also called weighted array)
    data_user_multiply = np.multiply(list_user_rating, data_user_genre_matrix_numpy)
    #print(data_user_multiply)
    
    #now making a weighted sumed matrix
    data_user_sum_weighted_matrix = np.sum(data_user_multiply, axis = 0) ##using np.sum() so as to get column wise sum
    #print(data_user_sum_weighted_matrix)
    
    #now making the ratings normalized
    normalized_data_user_sum_weighted_matrix = data_user_sum_weighted_matrix/sum(data_user_sum_weighted_matrix)
    #print(normalized_data_user_sum_weighted_matrix)
    
    """The previous step is done"""
    """Now by using normalized_data_user_sum_weighted_matrix, we will try to recommend the movies to user"""
    
    #making a new dataframe, where only genre scores are present
    datarecommedation_genre_frame = datarecommendation.drop(['movieId', 'title', 'rating'], axis=1)
    #print(datarecommedation_genre_frame)
    
    #making recommendation numpy array for multiplication
    data_recommendation_genre_matrix_numpy = np.array(datarecommedation_genre_frame)
    #print(data_recommendation_genre_matrix_numpy)
    
    #now making multiplied recommendation genre matrix
    data_recommendation_genre_matrix_multiply = np.multiply(normalized_data_user_sum_weighted_matrix, data_recommendation_genre_matrix_numpy)
    #print(data_recommendation_genre_matrix_multiply)
    
    #now making weighted recommendation matrix
    data_recommendation_weighted_matrix = np.sum(data_recommendation_genre_matrix_multiply, axis = 1)

    
    """Now since we got the recommendation matrix, now we will merge the matrix as a column in the datarecommendation dataframe"""
    datarecommendation['recommendation_rating'] = pd.Series(data_recommendation_weighted_matrix)
    #print(datarecommendation.head())
    datarecommendation = datarecommendation.sort_values('recommendation_rating', ascending = False)
    #print(datarecommendation.head())
    print(datarecommendation[['title', 'recommendation_rating']].head(10))
    #print(list(datarecommendation['title']))
    

In [77]:
recommendation_part1_CF(['toy story', 'heat', 'goldeneye', 'nixon', 'zootopia'], [4.5, 1.2, 2.6, 4.3, 4.9])

                        title  recommendation_rating
149                   hackers               0.641148
319  clear and present danger               0.612440
347                     speed               0.548644
415             fugitive, the               0.491228
83               nick of time               0.449761
70                  screamers               0.440191
102               taxi driver               0.413078
88               broken arrow               0.390750
80               white squall               0.390750
350                 true lies               0.390750


C:\Users\SHIVLEON\Anaconda3\lib\site-packages\pandas\core\indexing.py:1676: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)
C:\Users\SHIVLEON\Anaconda3\lib\site-packages\ipykernel_launcher.py:70: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


In [78]:
recommendation_part1_CF(['alien','alien from l.a.','alien nation','alien: covenant','alien: resurrection','aliens'], [4.3, 2.9, 3.4, 4.5, 4.6, 4.5])

                          title  recommendation_rating
88                 broken arrow               0.673442
323                   crow, the               0.551491
320                 client, the               0.551491
265     lã©on: the professional               0.551491
32                         babe               0.551491
141                crimson tide               0.551491
319    clear and present danger               0.550136
28   city of lost children, the               0.532520
326                forrest gump               0.510840
242         legends of the fall               0.510840


C:\Users\SHIVLEON\Anaconda3\lib\site-packages\pandas\core\indexing.py:1676: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self._setitem_single_column(ilocs[0], value, pi)
C:\Users\SHIVLEON\Anaconda3\lib\site-packages\ipykernel_launcher.py:70: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


In [79]:
list(movie_similarity_matrix_copy.columns)

["'71",
 "'burbs, the",
 "'night mother",
 "'round midnight",
 "'salem's lot",
 "'til there was you",
 '*batteries not included',
 '...and justice for all',
 '10',
 '10 cloverfield lane',
 '10 items or less',
 '10 rillington place',
 '10 things i hate about you',
 '10 to midnight',
 '10 years',
 '10,000 bc',
 '100 feet',
 '100 girls',
 '101 dalmatians',
 "101 dalmatians ii: patch's london adventure",
 '101 reykjavik',
 '102 dalmatians',
 '10th victim, the',
 '11\'09"01 - september 11',
 '11.22.63',
 '11:14',
 '12',
 '12 angry men',
 '12 chairs',
 '12 rounds',
 '12 strong',
 '12 years a slave',
 '127 hours',
 '12:01',
 '12:08 east of bucharest',
 '13',
 '13 assassins',
 '13 ghosts',
 '13 going on 30',
 '13 hours',
 '13 reasons why',
 '13 tzameti',
 '13th',
 '13th warrior, the',
 '1408',
 '1492: conquest of paradise',
 '15 minutes',
 '16 blocks',
 '17 again',
 '1776',
 '18 again!',
 '187',
 '1900',
 '1922',
 '1941',
 '1969',
 '1984',
 '2 days in new york',
 '2 days in paris',
 '2 days in